frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1264,6 +1264,37 @@ const IR::Node* TypeInference::postorder(IR::Type_Package* decl) {
 }
 
 const IR::Node* TypeInference::postorder(IR::Type_Specialized *type) {
+    // If there is no template_parameter in constructor parameters
+    // then compiler should reject instantiation
+    // because it cannot infer the type of template_parameter
+    for (auto arg : *(type->arguments)) {
+        if (arg->is<IR::Type_Dontcare>()) {
+            auto decl = refMap->getDeclaration(type->baseType->path);
+            if (auto extern_decl = decl->to<IR::Type_Extern>()) {
+                for (auto template_parameter : extern_decl->typeParameters->parameters) {
+                    for (auto meth : extern_decl->methods) {
+                        if (meth->type->returnType == nullptr) {
+                            bool found = false;
+                            for (auto meth_params : *(meth->getParameters())) {
+                                if (template_parameter->getP4Type()->equiv
+                                    (*(meth_params->type->getP4Type()))) {
```

Comment:
Why do the types need to be equivalent?
One of the types could have references to the other one, e.g., Generic<T>, where Generic is a generic type.

Reply:
The idea was to check if the template parameter is used somewhere between constructor parameters. 

The following code should be accepted (type can be infered from constructor parameter - but now I see that this isn't supported in the compiler yet - I might have to add that too) : 
```
extern E<T> {
    E(T t);
    void f(in T arg);
}
control c() {
    E<_>(0) e; 
    apply {
        e.f(0);
    }
} 
```
And the next one shouldn't (because function _f_ could be called with parameter of any type, for the same object e (just like with meters in original issue).: 
```
extern E<T> {
    E();
    void f(in T arg);
}
control c() {
    E<_>() e;       //rejected here
    apply {
        e.f(0);
        e.f("a"); 
    }
}
```

Reply:
For the first example, what type is used for the instantiation of `T`? If we knew the type of `0`, I'd agree, but the width has to be inferred too.

For the second example, I agree it shouldn't type check. But I'm not sure I agree with the sentence "function f could be called with parameter of any type" if you mean method `f` is polymorphic. It's not, because the type parameter is on the `extern` and not on the `f` method. So rather than thinking that "`f` could be called with parameter of any type," I would say it as "extern `E`'s type variable could be instantiated with any type, then method `f` would take an argument of the same type." Hope that makes sense?

Reply:
I printed the type of constructor argument and it was `int` for all integer arguments. `bit<N>` when width is specified, `Type_String` for strings etc...

```
extern E<T> {
    E(T t);
    void f(in T arg);
}
control c() {
    E<_>(0) e;      // Type: int
 // E<_>(8w10) e;   // Type: bit<8>
 // E<_>("a") e;    // Type: Type_String
    apply {
        e.f(0);
    }
} 
control proto();
package top(proto p);

top(c()) main;
```

I might have chosen the bad words in the explanation for second example. But I think we are on the same page - the type of argument that is used in constructor must be the only type that is used in method call.